Move the borders between qpu and qvm

mpharrigan opened the pull request:
A rough draft for what #356 might look like. 

 - `expectation` and `run_and_measure` have migrated to new `WavefunctionSimulator`
 - all the payload stuff and requests have migrated to `ForestConnection`. The user should never invoke methods on this object; rather it is composed into the following new objects:
 - `QPU`, `QVM`, and `WavefunctionSimulator` have appeared to take care of user niceties like sensible defaults and papering over the crazy expectation endpoint. They invoke the relevant methods on `ForestConnection`.
 - `run_and_measure` has disappeared from `QPU` and `QVM` because it is disrespectful towards quil and silently gives garbage results with noise.

## TODO

 - [x] review docstrings for copy-paste casualties
 - [ ] ~~helper methods like `get_qpu('agave')`, `get_qvm('noisy-9q-generic')` so the user doesn't have to construct their own `ForestConnection`~~ for follow-up PR
 - ~~[ ] give the same treatment to the compiler connection object~~
 - [x] Remove duplicated code from `QVMConnection` and `QPUConnection`. The goal is to keep these around for backwards compatibility.
 - [x] give `QPU` and `QVM` additional user-friendly functionality
 - [x] make the methods on `ForestConnection` private but make the constructor public

ntezak commented on pyquil/api/_base_connection.py:
```diff
@@ -180,3 +183,350 @@ def validate_run_items(run_items):
+    payload = {"type": TYPE_MULTISHOT_MEASURE,
+               "qubits": list(qubits),
+               "trials": trials,
+               'compiled-quil': quil_program.out()}
```
mega-nitpick: sneaky single quoted string (not incorrect but inconsistent)

ntezak commented on pyquil/api/_base_connection.py:
```diff
@@ -180,3 +183,350 @@ def validate_run_items(run_items):
+    else:
+        payload["compiled-quil"] = quil_program.out()
+
+    def _wrap_program(self, program):
```
I don't see how this could ever get called?

mpharrigan replied:
good spot

ntezak commented on pyquil/api/wavefunction_simulator.py:
```diff
@@ -0,0 +1,156 @@
+        """
+        Constructor for wavefunction simulation. Sets up any necessary security.
+
+        :param sync_endpoint: The endpoint of the server for running small jobs
```
These params are all wrong/copied pasted

mpharrigan replied:
updated docs for new classes `WavefunctionSimulator`, `QPU`, and `QVM`.

stevenheidel commented on pyquil/api/_qam.py:
```diff
@@ -0,0 +1,32 @@
+
+
+class QAM(ABC):
+    def run(self, quil_program: Program, classical_addresses: Iterable[int],
```
Should probably be annotated with @abstractmethod instead of raising

stevenheidel commented:
@mpharrigan - to help out the reviewers, could you post some code snippets of what running programs looks like before & after this PR?

mpharrigan commented:
We'll use this program

```python
from pyquil import Program
from pyquil.gates import *
prog = Program()
prog += H(0)
prog += CNOT(0, 1)
```

# The old way

```python
## Check out the wavefunction
# This creates an inconsistent API and abstraction
# between QVMConnection and QPUConnection
from pyquil import QVMConnection
import numpy as np
qvm = QVMConnection()
wfn = qvm.wavefunction(prog)
print(np.sum(np.abs(wfn.amplitudes) ** 2))

## Compute an expecatation
# Vestigial functions let you shoot yourself
# in the foot
from pyquil.paulis import sZ
eval = qvm.expectation(prog, [sZ(0)*sZ(1)]) # Oops! need to convert pauli's to programs
eval = qvm.pauli_expectation(prog, [sZ(0) * sZ(1)])
print(eval)

## Set up a qvm with a generic noise model
# This is very complicated and involves mocking out
# a "Device"
from pyquil.device import gates_in_isa, Device, ISA
from pyquil.noise import _decoherence_noise_model
class ShimDevice(Device):
    def __init__(self):
        isa = ISA.from_dict({
            '1Q':{
                '0':{},
                '1':{},
            },
            '2Q':{
                '0-1':{}
            }
        })
        gates = gates_in_isa(isa)
        self.noise_model = _decoherence_noise_model(gates)
noisy_qvm = QVMConnection(device=ShimDevice())

## Run under noise
# Functions silently produce bad results when
# run under noise
noisy_qvm.run_and_measure(prog, [0,1]) # Oops! doesn't work with noise model
noisy_qvm.pauli_expectation(prog, [sZ(0)*sZ(1)]) # Oops! doesn't work with noise model
prog += MEASURE(0,0)
prog += MEASURE(1,1)
bitstrings = np.asarray(noisy_qvm.run(prog, [0,1], 1000))

## Run under QPU
# The inconsistent API rears its head. Even though
# you might want to e.g. compute expectation values
# on a QPU using a sampling scheme.
from pyquil import QPUConnection
qpu = QPUConnection(device='8Q-Agave')
qpu.wavefunction(prog, [sZ(0)*sZ(1)]) # Oops! This method doesn't exist for QPU
qpu.expectation(prog, [sZ(0)*sZ(1)])  # Oops! This method doesn't exist for QPU
bitstrings = np.asarray(qpu.run(prog, [0,1], 1000))
```

# The new way

```python
## Check out the wavefunction
# now it's a different object so you know you're
# doing something unphysical
from pyquil.api import WavefunctionSimulator
import numpy as np
wfnsim = WavefunctionSimulator()
wfn = wfnsim.wavefunction(prog)
print(np.sum(np.abs(wfn.amplitudes) ** 2))

## Compute an expecatation
# Just keep the functions we use
from pyquil.paulis import sZ
eval = wfnsim.expectation(prog, [sZ(0)*sZ(1)])
print(eval)

## Set up a qvm with a generic noise model
# Cleaner, and now suitable for introduction of helper
# functions. Follow up work will introduce
# pre-configured QVM's and QPU's. 

def get_generic_4q_noisy_qvm():
    # TODO: introduce pre-configured `QuantumComputer`s like this.
    from pyquil.api import QuantumComputer, QVM, ForestConnection
    from pyquil.device import NxDevice
    import networkx as nx
    device = NxDevice(topology=nx.convert_node_labels_to_integers(nx.grid_2d_graph(2,2)))
    gates = gates_in_isa(device.get_isa())
    noise_model = _decoherence_noise_model(gates)
    return QuantumComputer(
        name='noisy-4q-qvm',
        qam=QVM(connection=ForestConnection(), noise_model=noise_model),
        device=device,
    )
noisy_qvm = get_generic_4q_noisy_qvm()
    

## Run under noise
# No longer can you enter a scenario where no noise is applied
noisy_qvm.run_and_measure(prog, [0,1], 1000) # Works now!
prog += MEASURE(0,0)
prog += MEASURE(1,1)
bitstrings = noisy_qvm.run(prog, [0,1], 1000) # numpy by default

## Run under QPU
def get_qpu(name):
    # TODO: introduce pre-configured `QuantumComputer`s like this
    from pyquil.api import get_devices
    device = get_devices()[name]
    from pyquil.api import QuantumComputer, QPU, ForestConnection
    return QuantumComputer(
        name=name,
        qam=QPU(connection=ForestConnection(), device=device),
        device=device,
    )
    
qpu = get_qpu('8Q-Agave')
qpu.expectation(prog, [sZ(0)*sZ(1)])  # Doesn't exist at present. Can be introduced
                                      # to compute via sampling and will automatically
                                      # work for both QPU and QVM.
bitstrings = qpu.run(prog, [0,1], 1000)
```

There's still some follow up work to be done: introduce helper functions for constructing `QuantumComputer` objects, more convenience methods for running programs like a sampling-based `expectation`, but this refactor makes the APIs consistent and the composition of objects amenable to helper functions and extensions

mpharrigan commented on pyquil/api/_base_connection.py:
```diff
@@ -180,3 +183,347 @@ def validate_run_items(run_items):
+def expectation_payload(prep_prog, operator_programs, random_seed):
+    """REST payload for :py:func:`expectation`"""
+    if operator_programs is None:
+        operator_programs = [Program()]
```
idk this was there in the original `QVMConnection` code

vtomole commented:
@mpharrigan Why a wavefunctionsimulator? QVM simulates quantum programs, so it makes sense to extract wavefunctions from it.

stevenheidel commented on pyquil/api/_qam.py:
```diff
@@ -0,0 +1,35 @@
+        :param int trials: Number of times to run the program (a.k.a. number of shots)
+        :return: A list of a list of classical registers (each register contains a bit)
+        """
+        pass
```
`pass` not needed when there's a comment

mpharrigan replied:
4a59dfacf41a106c8ddcf45d512c729acb9fb4cf

stevenheidel commented on pyquil/api/qpu.py:
```diff
@@ -247,6 +218,11 @@ def run_and_measure(self, quil_program, qubits, trials=1, needs_compilation=True
         :return: A list of a list of classical registers (each register contains a bit)
         :rtype: list
         """
+        # Developer note: Can't wholesale replace these functions with
```
I'm confused - we can't replace them because we _have_ deprecated or _have not_ deprecated the http apis?

mpharrigan replied:
clarified in 93e68fcd480a5884be1a87d2212661e643d24ce6

stevenheidel commented on pyquil/api/qpu.py:
```diff
@@ -306,15 +290,69 @@ def wait_for_job(self, job_id, ping_time=None, status_time=None):
+
+
+class QPU(QAM):
+    def __init__(self, connection: ForestConnection, device: Union[Device, str] = None):
```
Instead of exposing ForestConnection - why not create a `ForestOptions` dataclass which contains the set of options? I think that could get rid of a lot of the weirdness/duplication that results from needed to share a constructor

mpharrigan replied:
Let me think about this a bit

stevenheidel commented on pyquil/api/qpu.py:
```diff
@@ -306,15 +290,69 @@ def wait_for_job(self, job_id, ping_time=None, status_time=None):
+
+
+class QPU(QAM):
+    def __init__(self, connection: ForestConnection, device: Union[Device, str] = None):
```
device = None is a bad default since it will throw an error

mpharrigan replied:
good catch. fixed in 51a120cbd47182bc861cb7740d042b440d32ab7f

stevenheidel commented on pyquil/api/quantum_computer.py:
```diff
@@ -0,0 +1,183 @@
+    """
+    program = Program(program.instructions)  # Copy
+    to_measure = []
+    while True:
```
This won't work for empty programs or programs with only measure instructions (maybe you're okay with this)

mpharrigan replied:
good catch. fixed in 64f40391d02ede8b4c6972a0000dc1835869afcb

stevenheidel commented on pyquil/api/quantum_computer.py:
```diff
@@ -0,0 +1,183 @@
+
+        I think we should discourage use of this function. ``run`` is much more general.
+
+        Unlike the old ``run_and_measure`` function, this one works with noise models.
```
I'm fine with this as long as we include measure_all(). Then people could just do:
```
prog.measure_all()
qpu.run(prog)
# instead of
qpu.run_and_measure(prog)
```

mpharrigan replied:
changed my mind. `run_and_measure` is fine if it's just a convenience wrapping for adding measure and calling run (as it is now). The pain was in how the behavior differs between QVMConnection and QPUConnection (esp wrt noise models). Updated the docstrings

stevenheidel commented on pyquil/api/quantum_computer.py:
```diff
@@ -0,0 +1,183 @@
+    def get_isa(self, oneq_type='Xhalves', twoq_type='CZ'):
+        return self.device.get_isa(oneq_type=oneq_type, twoq_type=twoq_type)
+
+    def run(self, program, classical_addresses, trials, symmetrize=None) -> np.ndarray:
```
Are people going to be clear on the difference between QuantumComputer.run and QPU.run?

mpharrigan replied:
Probably not ... 

In my mind, QPU.run and QVM.run typify the `QAM` interface. This sets the minimal collection of methods that an abstract quantum machine needs to satisfy to work with pyquil. Currently it's just `run`. You can imagine slotting in additional QAM's like the referenceqvm work or other simulators. You can also imagine evolving the interface to have additional (but still "low level") methods.

`QuantumComputer` includes the qam, a description of the device (real or otherwise) and a bunch of convenience methods for executing quantum programs that build on top of `QAM.run`. If you add a new method (e.g. `sampling_expectation(prog, pauli)`) it instantly works with anything that satisfies the `QAM` interface. So there's no actual difference between `QPU.run` and `QuantumComputer.run`, but there *is* a difference between `QPU` and `QuantumComputer`

1) does this make any sense?
2) do you have suggestions for how to convey this to users?

stevenheidel commented on pyquil/device.py:
```diff
@@ -319,9 +323,56 @@ def from_dict(d):
+
+class AbstractDevice(ABC):
+
+    def qubit_topology(self) -> nx.Graph:
```
@abstractmethod instead of throwing NotImplemented

mpharrigan replied:
0d3d5e132dee1a2318eac31c5cb88a387e9a60e6

stevenheidel commented:
@mpharrigan - thanks for the detailed examples! Looks great overall. Also looking forward to the `get_qpu` et al functions

I'm in Toronto for the latter half of this week so if I'm a bottleneck at that point then feel free to just go ahead 👍 

Something else to keep in mind is what we're exposing at the top-level pyquil.__init__ - right now the example on this page: https://www.rigetti.com/products is 3 lines of imports when it could easily be 1 or 2

mpharrigan commented:
Thanks for the review!

> Something else to keep in mind is what we're exposing at the top-level pyquil.init - right now the example on this page: https://www.rigetti.com/products is 3 lines of imports when it could easily be 1 or 2

Totally on board with this. My gut is to keep `pyquil.gates` out of the top-level init file, so the example would have 2 lines of import

```python
from pyquil import Program, get_qc
from pyquil.gates import *
```

mpharrigan commented:
> I understand you're trying to build toward a scenario where users can more simply build ISAs and noise models, but I don't yet see that this is simple enough.

Point definitely taken that it isn't simple enough yet. I think this will be fixed with the helper constructor functions. Maybe I should bite the bullet and implement those in this PR.

>  In particular, in your example above you instantiate the QVM and the QPU with a noise model or a device, but then also supply a device to the QuantumComputer, to which the QVM/QPU is supplied. And conceptually, a device supplies more than just the ISA - it supplies the specs and the noise models, but that's all you have it supplying to the QVM.

I think you're sortof making my point. The device is a whole bunch of things: the ISA, specs, noise model, temperature, time of last retune, etc. None of those directly relate to the action of running quantum programs. To run a quantum program, the only info about the device you need is its name. In my example above I should have done `QPU(device=name)` to make this more clear.

When using a QVM, you can have a fake isa, specs, noise model, etc -- or you can use a "real" device -- because none of these things directly relate to the action of running quantum programs. To emulate the running of a quantum program, the only thing the QVM needs is a noise model.

mpharrigan commented:
> Point 3 above also leads me to ask about the NxDevice class, and whether you're aiming to remove specs and noise model from devices in general.

No, those can be there too. I've personally found the specs and per-device noise models less useful than topology/isa information but I'd like to hear from other users

ntezak commented:
@mpharrigan Then how about we at least rename `symmetrize` to `symmetrize_readout`? There are lots of other things one might symmetrize.

pkaralekas commented:
@mpharrigan there's a lot of code in here that I don't think necessarily has to do with moving the borders, specifically things like `NxDevice`

mpharrigan commented:
added tests

ntezak commented on pyquil/api/_qam.py:
```diff
@@ -0,0 +1,33 @@
+
+    @abstractmethod
+    def wait_for_job(self, job_id, ping_time=None, status_time=None):
+        pass
```
docstring? I know it's abstract, but for consistency?

ntezak commented on pyquil/api/qpu.py:
```diff
@@ -306,15 +279,62 @@ def wait_for_job(self, job_id, ping_time=None, status_time=None):
+                                              device_name=self.device_name)
+
+    def wait_for_job(self, job_id, ping_time=None, status_time=None) -> Job:
+        return self.connection._wait_for_job(job_id=job_id, ping_time=ping_time,
```
docstring

ntezak commented on pyquil/api/quantum_computer.py:
```diff
@@ -0,0 +1,187 @@
+            any MEASURE instructions you've added to your program
+        :param trials: The number of times to run the program.
+        :param symmetrize_readout: Whether to apply readout error symmetrization. If not specified,
+            the class attribute ``symmetrize_readout`` will be used. See
```
nitpick: Should this be "instance attribute"?

ntezak commented on pyquil/api/quantum_computer.py:
```diff
@@ -0,0 +1,187 @@
+        it will insert an ``X`` gate prior to any ``MEASURE`` instruction and then flip the
+        measured classical bit back.
+
+        See :py:func:`run` for this function's parameter descriptions.
```
I think it's slightly confusing to have two public methods`run(..., symmetrize_readout)` and `run_symmetrized_readout` with documentation shared between them (and one calling the other). As a user I would prefer a single public API call for this and a single docstring. My suggestion is to make this method private and include the documentation with either `run` or at the class toplevel.

mpharrigan replied:
I don't see how this is confusing. Why shouldn't one call the other?

ntezak replied:
Well, if I read the docs for `run` it points me to the docs for `run_symmetrized_readout`, which again refers me to the docs for `run`. So to fully understand either function you need to look at both docstrings in parallel.

mpharrigan replied:
I personally find that sort of cross-referenced documentation easier to grok than having one big wall of text that describes every possible way of running on the device

bear in mind that we'll probably be adding additional `run` variants in the future

ntezak replied:
Ok, works for me

ntezak commented on pyquil/api/qvm.py:
```diff
@@ -467,3 +449,87 @@ def _add_rng_seed_to_payload(self, payload):
+                                              random_seed=self.random_seed)
+
+    def wait_for_job(self, job_id, ping_time=None, status_time=None) -> Job:
+        return self.connection._wait_for_job(job_id=job_id, ping_time=ping_time,
```
docstring

ntezak commented on pyquil/api/wavefunction_simulator.py:
```diff
@@ -0,0 +1,149 @@
+                                                   classical_addresses=classical_addresses,
+                                                   random_seed=self.random_seed)
+
+    def expectation(self, prep_prog: Program, pauli_terms: Union[PauliSum, List[PauliTerm]]):
```
return type signature ` -> Union[float, List[float]]`

ntezak commented on pyquil/api/wavefunction_simulator.py:
```diff
@@ -0,0 +1,149 @@
+            return sum(results)
+        return results
+
+    def run_and_measure(self, quil_program: Program, qubits: List[int] = None, trials: int = 1):
```
return type signature

ntezak commented on pyquil/api/wavefunction_simulator.py:
```diff
@@ -0,0 +1,149 @@
+                                                      trials=trials, random_seed=self.random_seed)
+
+    def wait_for_job(self, job_id, ping_time=None, status_time=None) -> Job:
+        return self.connection._wait_for_job(job_id=job_id, ping_time=ping_time,
```
docstring

mpharrigan commented:
I added type signatures where it helped me / pycharm. Since types are optional in Python, there's more types than before, and pycharm can figure (most?) everything out, I don't think that should block this

ntezak commented:
Build is failing because numpy arrays implement shape broadcasting when using comparison operators. We can use `np.allclose` instead

mpharrigan commented:
fixed up the numpy issues so all the old functions will continue to return lists (with the exception of `job.result()` which now gives numpy arrays ... would be too kludgy to keep backwards compatible and advanced users of the async api's probably can deal with this difference / want numpy arrays anyway)

fixed style issues. fixed py35 test failures

Let's merge!
